packages/story-editor/src/components/canvas/frameElement.js

```diff
@@ -80,41 +90,72 @@ const EmptyFrame = styled.div`
 const NOOP = () => {};
 
 function FrameElement({ id }) {
-  const { isSelected, isBackground, element } = useStory(({ state }) => ({
-    isSelected: state.selectedElementIds.includes(id),
-    isBackground: state.currentPage?.elements[0].id === id,
-    element: state.currentPage?.elements.find((el) => el.id === id),
+  const [isTransforming, setIsTransforming] = useState(false);
+
+  const {
+    setNodeForElement,
+    handleSelectElement,
+    isEditing,
+    setEditingElement,
+    setEditingElementWithState,
+  } = useCanvas(({ state, actions }) => ({
+    setNodeForElement: actions.setNodeForElement,
+    handleSelectElement: actions.handleSelectElement,
+    isEditing: state.isEditing,
+    setEditingElement: actions.setEditingElement,
+    setEditingElementWithState: actions.setEditingElementWithState,
   }));
-  const setEditingElement = useCanvas(
-    ({ actions }) => actions.setEditingElement
-  );
+  const { isSelected, selectedElementIds, isActive, isBackground, element } =
+    useStory(({ state }) => ({
+      isSelected: state.selectedElementIds.includes(id),
+      isBackground: state.currentPage?.elements[0].id === id,
+      element: state.currentPage?.elements.find((el) => el.id === id),
+      selectedElementIds: state.selectedElementIds,
```

Comment:
This would already be a big improvement I think:

```js
  const { isSelected, selectedElementIds, isActive, isBackground, element } =
    useStory(({ state }) => {
      const isOnlySelectedElement = state.selectedElementIds.length === 1 &&
        state.selectedElementIds.includes(id);
      return {
        isSelected: state.selectedElementIds.includes(id),
        isBackground: state.currentPage?.elements[0].id === id,
        element: state.currentPage?.elements.find((el) => el.id === id),
        selectedElementIds: state.selectedElementIds,
        isOnlySelectedElement,
        isActive:
          isOnlySelectedElement
          !isTransforming &&
          !isEditing,
      }
    });
```

Then `TextFrame` just needs to check `isOnlySelectedElement`

Reply:
Refactored in https://github.com/GoogleForCreators/web-stories-wp/pull/10452/commits/be6a5baed59440a757780009a52b1fbc0c4ba63b

Reply:
Speaking in general, since we are breaking down the code into multiple packages, there would again be instances where we would want to use some states from `story-editor` ( for example we couldn't use `useStory` here ). The only option we have is that we use some workaround by passing down props to avoid their usage in another package ( Because other packages cannot import anything from `story-editor` to avoid circular dependencies ). This works for now, but it doesn't look very scalable and is probably not a proper solution.

💭  Maybe these are some more examples of why some of our states should live in some central package that can be shared among all packages ( https://github.com/GoogleForCreators/web-stories-wp/issues/10614 ), but not sure how these should be organized.